src/rate_limiter.rs

```diff
@@ -0,0 +1,247 @@
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::time::{Duration, Instant};
+
+use crate::TokenBucket;
+
+pub struct RateLimiter<'a, K> {
+    buckets: HashMap<K, TokenBucket<'a>>,
+    clock: &'a (dyn Fn() -> Instant + Sync),
+}
+
+impl<'a, K: Eq + Hash> RateLimiter<'a, K> {
```

Comment:
Nit. The "Rust Way" of implementing this would be to have two separate `impl` blocks: one for methods which require `K` to implement `Eq + Hash`, and one for the methods that don't. This is not strictly necessary, though.

Reply:
What's the benefit of the 'Rust Way'?

Reply:
This is not particular useful for `RateLimiter` (or `HashMap` for that matter), I agree, as the only thing you could for both is to create an "empty" instance. But you can imagine a case where you implement certain methods only if the type parameter allows that. E.g. https://doc.rust-lang.org/std/vec/struct.Vec.html#method.dedup is only implemented when the type of elements stored in such `Vec<T>` implements `PartialEq`.